### PR TITLE
#208 - Configurable Find Facets and Search Filters 

### PR DESCRIPTION
**Pull request by CalinL**
The goal of this feature is to implement the search filter configuration in the start page.
The admin can manage the filters for Products via the search configuration in Episerver> Start> Search settings.
The admin will be able to add, remove and move up/down the filters to reorder them.

### CODE REVIEW

File: src/Foundation.Find.Cms/Config/EnumHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+﻿using EPiServer.Framework.Localization;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web.Mvc;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/EnumHelpers.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/EnumSelectionDescriptionAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using System.ComponentModel;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/EnumSelectionDescriptionAttribute.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/EnumSelectionFactory.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+﻿using EPiServer.Shell.ObjectEditing;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/EnumSelectionFactory.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/FacetEnums.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
+﻿namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/FacetEnums.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/FacetFilterConfigurationItem.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
+﻿using EPiServer.DataAnnotations;
+using EPiServer.Globalization;
+using Foundation.Find.Cms.Facets;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/FacetFilterConfigurationItem.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/FacetFilterContentConfigurationItem.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/FacetFilterContentConfigurationItem.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/FacetFilterProductConfigurationItem.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/FacetFilterProductConfigurationItem.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/FacetFiltersConfigurationProperty.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿using EPiServer.PlugIn;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/FacetFiltersConfigurationProperty.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/IgnoreCollectionEditorDescriptor.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
+﻿using EPiServer.Cms.Shell.UI.ObjectEditing.EditorDescriptors;
+using EPiServer.DataAnnotations;
+using EPiServer.Shell.ObjectEditing;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/IgnoreCollectionEditorDescriptor.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/PropertyListBase.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+﻿using EPiServer.Core;
+using EPiServer.Framework.Serialization;
+using EPiServer.Framework.Serialization.Internal;
+using EPiServer.ServiceLocation;
+using Newtonsoft.Json;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/PropertyListBase.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

File: src/Foundation.Find.Cms/Config/SelectOneEnumAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
+﻿using EPiServer.Shell.ObjectEditing;
+using System;
+using System.Web.Mvc;
+
+namespace Foundation.Find.Cms.Config
```
Comment (lunchin) [inline src/Foundation.Find.Cms/Config/SelectOneEnumAttribute.cs]: Change namespace to Foundation.Find.Cms.Facets.Config

Comment (CalinL) [reply src/Foundation.Find.Cms/Config/SelectOneEnumAttribute.cs]: Done

File: src/Foundation.Demo/Initialize.cs [hunk 12]
```diff
@@ -206,5 +171,163 @@ private void OnSavingContent(object sender, ContentEventArgs contentEventArgs)
...
+        {
+            if (contentEventArgs.Content is DemoHomePage startPage)
+            {
+                InitFacetRegistry();
+            }
+        }
+
+        #region Initialize Facets
+
+        private void InitFacetRegistry()
```
[the remark this answers is not in the document]
Comment (CalinL) [reply src/Foundation.Demo/Initialize.cs]: Done. I have to use ServiceLocator.Current.GetInstance<IInitializationEngine> inside OnPublishedContent event.

File: src/Foundation.Demo/Initialize.cs [hunk 13]
```diff
@@ -206,5 +171,163 @@ private void OnSavingContent(object sender, ContentEventArgs contentEventArgs)
...
+        {
+            var registry = ServiceLocator.Current.GetInstance<IFacetRegistry>();
```
[the remark this answers is not in the document]
Comment (CalinL) [reply src/Foundation.Demo/Initialize.cs]: Done.

File: src/Foundation.Demo/Initialize.cs [hunk 14]
```diff
@@ -206,5 +171,163 @@ private void OnSavingContent(object sender, ContentEventArgs contentEventArgs)
...
+
+            registry.Initialize();
+
+            GetFacets().ForEach(x => registry.AddFacetDefinitions(x));
+        }
+
+        private List<FacetDefinition> GetFacets()
+        {
+            var contentLoader = ServiceLocator.Current.GetInstance<IContentLoader>();
```
[the remark this answers is not in the document]
Comment (CalinL) [reply src/Foundation.Demo/Initialize.cs]: Done

File: src/Foundation.Demo/Initialize.cs [hunk 15]
```diff
@@ -206,5 +171,163 @@ private void OnSavingContent(object sender, ContentEventArgs contentEventArgs)
...
+
+            var currentMarket = ServiceLocator.Current.GetInstance<ICurrentMarket>();
```
[the remark this answers is not in the document]
Comment (CalinL) [reply src/Foundation.Demo/Initialize.cs]: Done

File: src/Foundation.Demo/Initialize.cs [hunk 16]
```diff
@@ -206,5 +171,163 @@ private void OnSavingContent(object sender, ContentEventArgs contentEventArgs)
...
+        private void InitFacetRegistry()
+        {
+            var registry = ServiceLocator.Current.GetInstance<IFacetRegistry>();
+
+            registry.Initialize();
+
+            GetFacets().ForEach(x => registry.AddFacetDefinitions(x));
+        }
+
+        private List<FacetDefinition> GetFacets()
```
[the remark this answers is not in the document]
Comment (CalinL) [reply src/Foundation.Demo/Initialize.cs]: Done

File: src/Foundation.Find.Cms/Config/EnumHelpers.cs (hunk 1 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/EnumHelpers.cs]: Done

File: src/Foundation.Find.Cms/Config/EnumSelectionDescriptionAttribute.cs (hunk 2 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/EnumSelectionDescriptionAttribute.cs]: Done

File: src/Foundation.Find.Cms/Config/EnumSelectionFactory.cs (hunk 3 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/EnumSelectionFactory.cs]: Done

File: src/Foundation.Find.Cms/Config/FacetEnums.cs (hunk 4 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/FacetEnums.cs]: Done

File: src/Foundation.Find.Cms/Config/FacetFilterConfigurationItem.cs (hunk 5 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/FacetFilterConfigurationItem.cs]: Done

File: src/Foundation.Find.Cms/Config/FacetFilterContentConfigurationItem.cs (hunk 6 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/FacetFilterContentConfigurationItem.cs]: Done

File: src/Foundation.Find.Cms/Config/FacetFilterProductConfigurationItem.cs (hunk 7 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/FacetFilterProductConfigurationItem.cs]: Done

File: src/Foundation.Find.Cms/Config/FacetFiltersConfigurationProperty.cs (hunk 8 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/FacetFiltersConfigurationProperty.cs]: Done

File: src/Foundation.Find.Cms/Config/IgnoreCollectionEditorDescriptor.cs (hunk 9 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/IgnoreCollectionEditorDescriptor.cs]: Done

File: src/Foundation.Find.Cms/Config/PropertyListBase.cs (hunk 10 above)
Comment (CalinL) [reply src/Foundation.Find.Cms/Config/PropertyListBase.cs]: Done

**Pull request merged**